server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java

```diff
@@ -334,4 +360,113 @@ private CircuitBreaker validateAndCreateBreaker(BreakerSettings breakerSettings)
+        private final long minimumInterval;
+
+        private long blackHole;
+        private final Object lock = new Object();
+
+        G1DoubleCheckStrategy(JvmInfo jvmInfo, LongSupplier currentMemoryUsageSupplier,
+                                     LongSupplier timeSupplier, long minimumInterval) {
+            assert minimumInterval > 0;
+            this.currentMemoryUsageSupplier = currentMemoryUsageSupplier;
+            this.timeSupplier = timeSupplier;
+            this.minimumInterval = minimumInterval;
+            long g1RegionSize = jvmInfo.getG1RegionSize();
+            if (g1RegionSize <= 0) {
+
+                this.g1RegionSize = fallbackRegionSize(jvmInfo);
+            } else {
+                this.g1RegionSize = g1RegionSize;
+            }
+        }
+
+        static long fallbackRegionSize(JvmInfo jvmInfo) {
+            // mimick JDK calculation
+            long averageHeapSize =
+                (jvmInfo.getMem().getHeapMax().getBytes() + JvmInfo.jvmInfo().getMem().getHeapMax().getBytes()) / 2;
+            long regionSize = Long.highestOneBit(averageHeapSize / 2048);
+            if (regionSize < ByteSizeUnit.MB.toBytes(1)) {
+                regionSize = ByteSizeUnit.MB.toBytes(1);
+            } else if (regionSize > ByteSizeUnit.MB.toBytes(32)) {
+                regionSize = ByteSizeUnit.MB.toBytes(32);
+            }
+            return regionSize;
+        }
+
+        @Override
+        public MemoryUsage doubleCheck(MemoryUsage memoryUsed) {
+            long maxHeap = JvmInfo.jvmInfo().getMem().getHeapMax().getBytes();
+            boolean leader;
+            synchronized (lock) {
+                leader = timeSupplier.getAsLong() >= lastCheckTime + minimumInterval;
+                doubleCheckingRealMemoryUsed(leader);
+                if (leader) {
+                    logger.info("attempting to trigger G1GC due to high heap usage [{}]", memoryUsed.baseUsage);
+                    long localBlackHole = 0;
+                    // number of allocations, corresponding to (approximately) number of free regions + 1
+                    long allocationCount = (maxHeap - memoryUsed.baseUsage) / g1RegionSize + 1;
+                    // allocations of half-region size becomes single humongous alloc, thus taking up a full region.
+                    int allocationSize = (int) (g1RegionSize >> 1);
+                    long maxUsageObserved = memoryUsed.baseUsage;
+                    for (long i = 0; i < allocationCount; ++i) {
+                        long current = currentMemoryUsageSupplier.getAsLong();
+                        if (current >= maxUsageObserved) {
+                            maxUsageObserved = current;
+                        } else {
+                            // we observed a memory drop, so some GC must have occurred
+                            break;
+                        }
+                        localBlackHole += new byte[allocationSize].hashCode();
```

Comment:
Yes and no.

In theory yes, if there is really no collectible heap left.

But if that was the case, just creating the `CircuitBreakingException` poses the same risk. And if we are that close, we are doomed anyway I think. The chances of us having a workload at exactly 99.95 percent heap  (corresponding to approximately 2000 regions) and surviving is so small that even if it was the case, the next time round we enter the same workload it would fall over.

Notice that we only need 1 region of space free or collectible space for this to succeed.